bin/varnishncsa/varnishncsa.c

```diff
@@ -160,8 +160,26 @@ static struct ctx {
 	const char		*handling;
 	const char		*side;
 	int64_t			vxid;
+	int			vcl_recv_seen;
+	int			vcl_br_seen;
 } CTX;
 
+enum update_mode {
+	UM_NEVER,
+	UM_IF_NOT_SET,
+	UM_ALWAYS
+};
+
+#define UPDATE_REQ_C(CTX) (CTX.vcl_recv_seen ? UM_NEVER : UM_ALWAYS)
+#define UPDATE_REQ_B(CTX) (CTX.vcl_br_seen ? UM_NEVER : UM_ALWAYS)
+
+#define UPDATE_REQ(CTX) (*CTX.side == 'b' ? UPDATE_REQ_B(CTX) : UPDATE_REQ_C(CTX))
+
+#define UPDATE_RESP_C(CTX) (UM_ALWAYS)
+#define UPDATE_RESP_B(CTX) (CTX.vcl_br_seen ? UM_NEVER : UM_ALWAYS)
+
+#define UPDATE_RESP(CTX) (*CTX.side == 'b' ? UPDATE_RESP_B(CTX) : UPDATE_RESP_C(CTX))
```

Comment:
I think I would prefer something like this:

```c
enum format_policy {
        FMTPOL_FIRST,
        FMTPOL_REQ,
        FMTPOL_RESP,
        FMTPOL_BEREQ,
        FMTPOL_BERESP,
};
```

Remember that you always have access to `CTX` to apply the policy, so we shouldn't need macros. A single function could tell you what to do:

```c
static unsigned
process_format(enum format_policy fp)
{

        /* switch (fp) => check CTX */
}
```

Then you can use that in functions like `frag_fields()` to bail out early:

```c
if (!process_format(fp))
        return;
```

It might prevent cascading enum arguments, I could only have a quick look at the patch series.

Reply:
I have force pushed an updated version that includes your suggested changes.